autokeras/hypermodel/graph.py

```diff
@@ -463,4 +476,69 @@ def hyper_build(self, hp):
         outputs = []
         for output_node in self.outputs:
             outputs.append(old_node_to_new[output_node])
-        return PlainGraph(inputs, outputs, override_hps=self.override_hps)
+        return PlainGraph(inputs=inputs,
+                          outputs=outputs,
+                          override_hps=self.override_hps)
+
+    def get_config(self):
+        config = super().get_config()
+        block_classes = [block.__class__.__name__ for block in self.blocks]
+        node_classes = [node.__class__.__name__ for node in self._nodes]
+        block_inputs = {
+            str(block_id): [self._node_to_id[node]
+                            for node in block.inputs]
+            for block_id, block in enumerate(self.blocks)}
+        block_outputs = {
+            str(block_id): [self._node_to_id[node]
+                            for node in block.outputs]
+            for block_id, block in enumerate(self.blocks)}
+        node_inputs = {
+            str(node_id): [self._block_to_id[block]
+                           for block in node.in_blocks]
+            for node_id, node in enumerate(self._nodes)}
+        node_outputs = {
+            str(node_id): [self._block_to_id[block]
+                           for block in node.out_blocks]
+            for node_id, node in enumerate(self._nodes)}
+
+        inputs = [self._node_to_id[node] for node in self.inputs]
+        outputs = [self._node_to_id[node] for node in self.outputs]
+
+        config.update({
+            'inputs': inputs,  # List of node_ids.
+            'outputs': outputs,  # List of node_ids.
+            'block_classes': block_classes,  # List of strings of class names.
+            'node_classes': node_classes,  # List of strings of class names.
+            'block_inputs': block_inputs,  # Dict {id: List of node_ids}.
+            'block_outputs': block_outputs,  # Dict {id: List of node_ids}.
+            'node_inputs': node_inputs,  # Dict {id: List of block_ids}.
+            'node_outputs': node_outputs,  # Dict {id: List of block_ids}.
+        })
+        return config
+
+    def set_config(self, config):
+        self.blocks = [getattr(autokeras, block_class)()
```

Comment:
I expect you'll want to support user-defined blocks

To do so, if blocks have a `from_config` like described above, then you can make use of Keras's ` custom_object` mechanisms:

https://github.com/tensorflow/tensorflow/blob/649d1b34da7a90f0dbc024911751a34bef591290/tensorflow/python/keras/utils/generic_utils.py#L42

```python
@classmethod
def from_config(self, config, custom_objects=None):
  self.blocks = [
     tf.keras.deserialize_keras_object(block_config, custom_objects=custom_objects)
    for block_config in config['blocks']]
...
```

Reply:
Same for nodes below

Reply:
To my understanding, all the blocks in AutoKeras are custom_objects to Keras. So we need to union the AutoKeras objects dict and the custom_objects passed by the user as the `custom_objects` to the `deserialize_keras_object` function, right?

Reply:
That's one approach, another way is to do a module-level approach like what Keras does, which is that for each object type, there is a `deserialize` method (for example, `tf.keras.metrics.deserialize(config, custom_objects=None)`):

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/metrics.py#L3214

This function collects that particular module's `globals()` and uses those as the deserialization names and objects dict.

So you could have something like `autokeras.blocks.deserialize(config, custom_objects=None)` that hooks into `deserialize_keras_object` in the same way. Then you would use that when you know the thing you're deserializing is a Block